crates/testing/src/byzantine/byzantine_behaviour.rs

```diff
@@ -0,0 +1,314 @@
+                            view_number: view_num,
+                            cached_event: event.clone(),
+                        });
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    fn compare_views(current_view: u64, cached_view: u64, number_of_views_to_delay: u64) -> bool {
+        if current_view <= cached_view {
+            return false;
+        }
+
+        if current_view - cached_view > number_of_views_to_delay {
+            return true;
+        }
+
+        false
+    }
+
+    async fn handle_event(&mut self, event: &HotShotEvent<TYPES>) -> HotShotEvent<TYPES> {
+        if let Some(event_info) = self.extract_hotshot_event_info(event) {
+            // Are we done with the view delay
+            if event_info.view_number > self.stop_view_delay_at_view_number {
+                return event.clone();
+            }
+
+            match self.received_events.entry(event_info.event_name.clone()) {
+                Entry::Occupied(mut event_info_entry) => {
+                    let cached_event_info = event_info_entry.get_mut();
+                    let cached_event = cached_event_info.cached_event.clone();
+
+                    match event {
+                        HotShotEvent::QuorumVoteRecv(_) | HotShotEvent::DaVoteRecv(_) => {
+                            // Ensure we delay all votes for a view
+                            if Self::compare_views(
+                                event_info.view_number,
+                                cached_event_info.view_number,
+                                self.number_of_views_to_delay,
+                            ) {
+                                let votes_received_count = self
+                                    .vote_rcv_count
+                                    .entry(event_info.event_name.clone())
+                                    .or_insert(0);
+                                *votes_received_count += 1;
+                                if *votes_received_count == self.num_nodes_with_stake - 1 {
+                                    *cached_event_info = event_info;
+                                    *votes_received_count = 0;
+                                }
+                            }
+                        }
+                        _ => {
+                            if Self::compare_views(
+                                event_info.view_number,
+                                cached_event_info.view_number,
+                                self.number_of_views_to_delay,
+                            ) {
+                                *cached_event_info = event_info;
+                            }
+                        }
+                    }
+                    return cached_event;
+                }
+                Entry::Vacant(vacant) => {
+                    vacant.insert(event_info);
+                }
+            }
+        }
+        event.clone()
+    }
```

Comment:
Something feels off here.  I think we basically just storing all events from previous views and then sending them after we've delayed enough views.  Why do we have to do all this logic with the votes?  

Maybe I'm missing something but I feel like the logic can simply be, have a map from view to vec of events.  if the event has a view number push it to the vec for that view, return all events from the view - num views to delay.  If it has no view number, either add it to the highest views vec, or just return that event.

If we did that we can get rid of `HotShotEventInfo` (which I think we should do anyway)